pkg/network/setup/netconf.go

```diff
@@ -0,0 +1,82 @@
+/*
+ * This file is part of the KubeVirt project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright 2021 Red Hat, Inc.
+ *
+ */
+
+package network
+
+import (
+	"fmt"
+	"sync"
+
+	v1 "kubevirt.io/api/core/v1"
+	"kubevirt.io/kubevirt/pkg/network/cache"
+)
+
+type NetConf struct {
+	setupCompleted    sync.Map
+	ifaceCacheFactory cache.InterfaceCacheFactory
+}
+
+func NewNetConf(ifaceCacheFactory cache.InterfaceCacheFactory) *NetConf {
+	return &NetConf{
+		setupCompleted:    sync.Map{},
+		ifaceCacheFactory: ifaceCacheFactory,
+	}
+}
+
+// Setup applies (privilege) network related changes for an existing virt-launcher pod.
+// As the changes are performed in the virt-launcher network namespace, which is relative expensive,
+// an early cache check is performed to avoid executing the same operation again (if the last one completed).
+func (c *NetConf) Setup(vmi *v1.VirtualMachineInstance, launcherPid int, doNetNS func(func() error) error, preSetup func() error) error {
+	id := vmi.UID
+	if _, exists := c.setupCompleted.Load(id); exists {
+		return nil
+	}
+
+	if err := preSetup(); err != nil {
+		return fmt.Errorf("setup failed, err: %w", err)
```

Comment:
@EdDev WDYT about changing this error message to "pre-setup failed" in order to distinguish it from the error message  of the actual setup?

Reply:
Usually my thumb rule about errors is this: Each level should throw the information relevant to its abstraction level and expect the same from the errors it needs to treat.
At this abstraction level, the added value is about the method of `Setup` which failed and the details are expected to come from the error that is attached.

Should we add the info about it being "pre-setup"? I have mixed thoughts about it... On one hand, as it is a function pointer, the caller may not have included this info and we cannot assume it. On the other hand, this does not work well with the error message "best practice" I mentioned earlier.

I tend to do a mix here: "setup failed at pre-setup stage, err: ...". WDYT?

Reply:
I'm with you on "setup failed at pre-setup stage, err: ...", sounds like a good compromise